### ISSUE TITLE
floor/ceiling of odd or even number divided by 2 could be simpler

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#24599]
An issue raised on [SO](https://stackoverflow.com/questions/75236716/anyway-to-get-rid-of-math-floor-for-positive-odd-integers-with-sympy-simplify/75241386#75241386) brings up the question of whether the simple case of `odd` or `even` divided by 2 should automatically return the equivalent non-function form of the expression: `(odd - 1)/2` or `even/2`.

Also, the expression `(odd+/-1)/2` should be recognized as an integer since `odd +/- 1` is recognized as even and `even/2` is recognized as being an integer:
```python
>>> o, e = Dummy(odd=True), Dummy(even=True)
>>> (e/2).is_integer
True
>>> (o+1).is_even
True
>>> ((o+1)/2).is_integer
```

**Comment by asmeurer** [comment]
I don't think it automatically should (I want to see less automatic evaluation based on assumptions), but it's definitely something that should simplify somewhere.

**Comment by oscarbenjamin** [comment]
> it's definitely something that should simplify somewhere.

The problem is that right now there isn't an effective place to make it happen unless it's the core assumptions but then it slows everything down.

**Comment by smichr** [comment]
`floor(even/2)->even/2` already. Is there a good reason to allow that but not the odd case (to `(odd-1)/2`)? Or do you think the even case should not be evaluating, too?

**Comment by asmeurer** [comment]
`doit` and `_eval_simplify` are both reasonable catchalls for things like this that don't fit into an existing simplification function (like `rewrite`, or `expand`). In the custom functions guide I recommend using `doit` for things that you want to put in `eval` but really shouldn't be there. https://docs.sympy.org/latest/guides/custom-functions.html#rewriting-and-simplification

**Comment by smichr** [comment]
I would really like to see `expand(basic=True)` do this. At last look, I don't think there is much use of that flag.

I don't like `doit` any more than `simplify` -- if you run `doit` on an expression, there might be lots of things attempted. I suppose one could do a replace(lambda x: x.func is floor, lambda x: x.doit(deep=False))`.

**Comment by asmeurer** [comment]
> Or do you think the even case should not be evaluating, too?

I don't think any function should do evaluation based on assumptions. As soon as you start writing `is_even` or `is_real` or whatever in the *`eval`* method, you run into performance problems. I think `eval` should only evaluate things on explicit numeric arguments, or some reasonable extension of this (like "an explicit integer multiple of π" in trig functions). The equivalent evaluation on symbolic arguments (i.e., using assumptions) should be done in `doit` or somewhere else, so that it can be done.

The problem when you have this automatic evaluation is that just creating an expression goes down a huge rabbit hole of looking up assumptions, and it completely kills performance. Just look at how long it takes to create the expression for `cos(pi/257).rewrite(sqrt)` (see the discussion at https://github.com/sympy/sympy/issues/24565). That's just *creating* the expression, which takes over 25 seconds, because it does a ton of evaluation. The expression is large but not *that* large, and SymPy ought to be able to create it instantly (and indeed it can, if you create it with `evaluate=False`).

**Comment by oscarbenjamin** [comment]
> Is there a good reason

Any logic in evaluation or in the assumptions slows everything down. There is always a good reason not to add more code there.

**Comment by asmeurer** [comment]
`expand(basic=True)` is another option too I guess. To me `doit` makes the most sense as specifically an "evaluate" for things that would be done in `eval` if the performance weren't an issue. The way I think of it is that you have some complex evaluation logic in `doit` that uses assumptions, and you apply the same logic in `eval` but only when the arguments are explicit numbers. In principle one could just make `eval` default to just calling `doit` iff the arguments are all `Number`, much like it already defaults to calling `evalf` if the arguments are all `Float`. 

Also `doit` allows arbitrary keyword arguments, so if a lot of things end up going there you can in principle start to control what happens with different flags. 

To be sure, `doit` itself could use some cleaning up so that more of the logic is handled in Basic rather than having to be reimplemented in each class https://github.com/sympy/sympy/issues/23609. That's one way in which `_eval_expand_basic` is currently better.

**Comment by sylee957** [comment]
The problem is that the assumptions from symbols could be better distinguished, 
because it is already cached in some way and should be simple.

**Comment by smichr** [comment]
All good points. I think the lure initially was resolving assumptions for simple expressions...like `odd + 1` being even. But if you handle that, then why not 3 args, etc... I can now totally agree with keeping this separate with benefit of hindsight.

Fortunately we can use workaround (like replacing `odd` with `2*int + 1` or using `replace` and explicitly testing assumptions).

**Comment by oscarbenjamin** [comment]
In this case it is actually evaluation that prevents the assumption from being obvious:
```
In [6]: Mul(o + 1, S.Half, evaluate=False).is_integer
Out[6]: True
```
The problem is that what you actually have is not `(o + 1)/2` but `o/2 + 1/2`. So now to determine if this is an integer you need to start counting half-integer terms rather than just checking if you have a sum of integers.
